### DISCUSSION TITLE
Nullref when writing to GITHUB_ENV from python

### DISCUSSION THREAD

**Discussion by jbzdarkid** [community/community#25753] [Actions]
<p>I’m somehow getting the classic C# null reference when trying to write to GITHUB_ENV.</p>
<pre><code class="lang-auto">import environ
with open(environ['GITHUB_ENV'], 'a') as f:
  subcomment = """\
Please verify the following diffs:
- [ ] All articles ran in 0:00:56.697371: [en](https://wiki.tf/d/3106540) [ar](https://wiki.tf/d/3106518) [cs](https://wiki.tf/d/3106519) [da](https://wiki.tf/d/3106520) [de](https://wiki.tf/d/3106521) [es](https://wiki.tf/d/3106522) [fi](https://wiki.tf/d/3106523) [fr](https://wiki.tf/d/3106524) [hu](https://wiki.tf/d/3106525) [it](https://wiki.tf/d/3106526) [ja](https://wiki.tf/d/3106527) [ko](https://wiki.tf/d/3106528) [nl](https://wiki.tf/d/3106529) [no](https://wiki.tf/d/3106530) [pl](https://wiki.tf/d/3106531) [pt](https://wiki.tf/d/3106532) [pt-br](https://wiki.tf/d/3106533) [ro](https://wiki.tf/d/3106534) [ru](https://wiki.tf/d/3106535) [sv](https://wiki.tf/d/3106536) [tr](https://wiki.tf/d/3106537) [zh-hans](https://wiki.tf/d/3106538) [zh-hant](https://wiki.tf/d/3106539)
- [ ] Wanted templates ran in 0:00:37.264911: [en](https://wiki.tf/d/3106541)"""
  f.write('GITHUB_COMMENT&lt;&lt;EOF')
  f.write(subcomment)
  f.write('EOF')
</code></pre>
<pre><code class="lang-auto">2022-01-06T15:26:24.4627453Z ##[error]Unable to process file command 'env' successfully.
2022-01-06T15:26:24.4642131Z ##[error]Object reference not set to an instance of an object.
</code></pre>
<aside class="onebox githubactions" data-onebox-src="https://github.com/jbzdarkid/TFWiki-scripts/actions/runs/1663629178">
  <header class="source">

      <a href="https://github.com/jbzdarkid/TFWiki-scripts/actions/runs/1663629178" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>

  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Commit">

      <svg width="24" height="24" class="github-icon github-icon-failure" viewbox="0 0 24 24" aria-hidden="true"><path d="M1 12C1 5.925 5.925 1 12 1s11 4.925 11 11-4.925 11-11 11S1 18.075 1 12zm8.036-4.024a.75.75 0 00-1.06 1.06L10.939 12l-2.963 2.963a.75.75 0 101.06 1.06L12 13.06l2.963 2.964a.75.75 0 001.061-1.06L13.061 12l2.963-2.964a.75.75 0 10-1.06-1.06L12 10.939 9.036 7.976z"></path></svg>

  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/jbzdarkid/TFWiki-scripts/actions/runs/1663629178" target="_blank" rel="noopener nofollow ugc">minimal repro, I guess</a>
    </h4>

      <div class="github-info">
        TFWiki Stats <span class="github-run-number">#134</span>
      </div>
  </div>
</div>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>This doesn’t seem to repro when I create the variable inside the .yml, though I might be doing it wrong.</p>

**Comment by airtower-luna** [comment, marked as answer]
<p><code>f.write()</code> writes exactly the data you give it, so your data is lacking the required newlines. Just add them:</p>
<pre><code class="lang-auto">  f.write('GITHUB_COMMENT&lt;&lt;EOF\n')
  f.write(subcomment)
  f.write('\nEOF\n')
</code></pre>
<p>Writing a newline before <code>EOF</code> guarantees it always appears on its own line, you can skip that if the <code>subcomment</code> always ends with one.</p>

**Comment by jbzdarkid** [comment]
<p>That worked, thank you!<br>
I’m still not sure why it’s throwing a nullref, but I guess I don’t care :slight_smile:</p>

**Comment by jsoref** [comment]
<p>Your python isn’t directly throwing that error. It’s constructing a malformed file which the .net runner is crashing on.</p>
<p>You could file a bug about the unhelpful error reporting here:</p><aside class="onebox allowlistedgeneric" data-onebox-src="https://github.com/actions/runner">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">

      <a href="https://github.com/actions/runner" target="_blank" rel="noopener nofollow ugc">GitHub</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/345;"><img src="https://user-images.githubusercontent.com/2119212/181093052-d67508a2-2bb6-4366-8ce5-31597591a735.png" class="thumbnail" width="690" height="345" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/4/a/4ac3c934860f292384221b7fd054bc4f3533b795_2_690x345.png, https://aws1.discourse-cdn.com/github/optimized/3X/4/a/4ac3c934860f292384221b7fd054bc4f3533b795_2_1035x517.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/4/a/4ac3c934860f292384221b7fd054bc4f3533b795.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/4/a/4ac3c934860f292384221b7fd054bc4f3533b795_2_10x10.png"></div>

<h3><a href="https://github.com/actions/runner" target="_blank" rel="noopener nofollow ugc">GitHub - actions/runner: The Runner for GitHub Actions</a></h3>

  <p>The Runner for GitHub Actions :rocket:. Contribute to actions/runner development by creating an account on GitHub.</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

**Comment by jbzdarkid** [comment]
<aside class="onebox githubissue" data-onebox-src="https://github.com/actions/runner/issues/1596">
  <header class="source">

      <a href="https://github.com/actions/runner/issues/1596" target="_blank" rel="noopener nofollow ugc">github.com/actions/runner</a>
  </header>

  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Issue">
	  <svg width="60" height="60" class="github-icon" viewbox="0 0 14 16" aria-hidden="true"><path d="M7 2.3c3.14 0 5.7 2.56 5.7 5.7s-2.56 5.7-5.7 5.7A5.71 5.71 0 0 1 1.3 8c0-3.14 2.56-5.7 5.7-5.7zM7 1C3.14 1 0 4.14 0 8s3.14 7 7 7 7-3.14 7-7-3.14-7-7-7zm1 3H6v5h2V4zm0 6H6v2h2v-2z"></path></svg>
  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/actions/runner/issues/1596" target="_blank" rel="noopener nofollow ugc">Unhelpful error message when writing malformed data to $GITHUB_ENV</a>
    </h4>

    <div class="github-info">
      <div class="date">
        opened <span class="discourse-local-date" data-format="ll" data-date="2022-01-07" data-time="15:41:08" data-timezone="UTC">03:41PM - 07 Jan 22 UTC</span>
      </div>


      <div class="user">
        <a href="https://github.com/jbzdarkid" target="_blank" rel="noopener nofollow ugc">
          <img alt="jbzdarkid" src="https://user-images.githubusercontent.com/5459085/181093054-a64c2bc8-596e-4177-8db2-20b4b684c5a0.jpeg" class="onebox-avatar-inline" width="20" height="20">
          jbzdarkid
        </a>
      </div>
    </div>

    <div class="labels">
        <span style="display:inline-block;margin-top:2px;background-color: #B8B8B8;padding: 2px;border-radius: 4px;color: #fff;margin-left: 3px;">
          bug
        </span>
    </div>
  </div>
</div>

  <div class="github-row">
    <p class="github-body-container">&lt;!--
👋 You're opening a bug report against the GitHub Actions **runner applicat<span class="show-more-container"><a href="" rel="noopener" class="show-more">…</a></span><span class="excerpt hidden">ion**.

🛑 Please stop if you're not certain that the bug you're seeing is in the runner application - if you have general problems with actions, workflows, or runners, please see the [GitHub Community Support Forum](https://github.community/c/code-to-cloud/52) which is actively monitored.  Using the forum ensures that we route your problem to the correct team.  😃
--&gt;

**Describe the bug**
When writing a particular string to `$GITHUB_ENV`, the runner throws a nullref, i.e. `Object reference not set to an instance of an object.`

**To Reproduce**
See https://github.community/t/nullref-when-writing-to-github-env-from-python/221606 . The following python code reproduces the issue.
```py
import environ
with open(environ['GITHUB_ENV'], 'a') as f:
  subcomment = """\
Please verify the following diffs:
- [ ] All articles ran in 0:00:56.697371: [en](https://wiki.tf/d/3106540) [ar](https://wiki.tf/d/3106518) [cs](https://wiki.tf/d/3106519) [da](https://wiki.tf/d/3106520) [de](https://wiki.tf/d/3106521) [es](https://wiki.tf/d/3106522) [fi](https://wiki.tf/d/3106523) [fr](https://wiki.tf/d/3106524) [hu](https://wiki.tf/d/3106525) [it](https://wiki.tf/d/3106526) [ja](https://wiki.tf/d/3106527) [ko](https://wiki.tf/d/3106528) [nl](https://wiki.tf/d/3106529) [no](https://wiki.tf/d/3106530) [pl](https://wiki.tf/d/3106531) [pt](https://wiki.tf/d/3106532) [pt-br](https://wiki.tf/d/3106533) [ro](https://wiki.tf/d/3106534) [ru](https://wiki.tf/d/3106535) [sv](https://wiki.tf/d/3106536) [tr](https://wiki.tf/d/3106537) [zh-hans](https://wiki.tf/d/3106538) [zh-hant](https://wiki.tf/d/3106539)
- [ ] Wanted templates ran in 0:00:37.264911: [en](https://wiki.tf/d/3106541)"""
  f.write('GITHUB_COMMENT&lt;&lt;EOF')
  f.write(subcomment)
  f.write('EOF')
```

**Expected behavior**
The bug here is about a lack of newline after the EOF marker. So, I would expect an error which says something like "malformed EOF marker", or "EOF marker too long". A very precient error might say "EOF marker missing newline".

## Runner Version and Platform
Runner version 2.285.1, Ubuntu 20.04.3

## What's not working?
https://github.com/jbzdarkid/TFWiki-scripts/runs/4728629242?check_suite_focus=true

## Job Log Output
```
2022-01-06T15:26:24.2881206Z ##[group]Run python master.py pull_request
2022-01-06T15:26:24.2882292Z [36;1mpython master.py pull_request[0m
2022-01-06T15:26:24.2930406Z shell: /usr/bin/bash -e {0}
2022-01-06T15:26:24.2930893Z env:
2022-01-06T15:26:24.2931483Z   pythonLocation: /opt/hostedtoolcache/Python/3.8.12/x64
2022-01-06T15:26:24.2932252Z   LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.8.12/x64/lib
2022-01-06T15:26:24.2933714Z   WIKI_USERNAME: ***
2022-01-06T15:26:24.2934234Z   WIKI_PASSWORD: ***
2022-01-06T15:26:24.2934678Z ##[endgroup]
2022-01-06T15:26:24.4627453Z ##[error]Unable to process file command 'env' successfully.
2022-01-06T15:26:24.4642131Z ##[error]Object reference not set to an instance of an object.
2022-01-06T15:26:24.4820362Z ##[group]Run peter-evans/find-comment@v1
2022-01-06T15:26:24.4820930Z with:
2022-01-06T15:26:24.4821305Z   issue-number: 13
2022-01-06T15:26:24.4821857Z   comment-author: github-actions[bot]
2022-01-06T15:26:24.4822902Z   token: ***
2022-01-06T15:26:24.4823414Z   repository: jbzdarkid/TFWiki-scripts
2022-01-06T15:26:24.4823954Z   direction: first
2022-01-06T15:26:24.4824355Z env:
2022-01-06T15:26:24.4824913Z   pythonLocation: /opt/hostedtoolcache/Python/3.8.12/x64
2022-01-06T15:26:24.4825643Z   LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.8.12/x64/lib
2022-01-06T15:26:24.4826209Z ##[endgroup]
```</span></p>
  </div>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>
